File: leet150/src/sliding_windown/best_time_to_buy_and_sell_stock_121_easy.rs
pub struct Solution;


impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        
        // so for sliding window we have left and right
        // right alwasy moves (seeking)
        // left just moves if the condition is unsastified
        let prices_len = prices.len();
        let mut max_profit = 0;
        let mut cheapest = prices[0];

        for right in 0..prices_len{

            if prices[right] < cheapest{
                cheapest = prices[right];
            }

            let profit = prices[right] - cheapest;
            // if profit for the round is bigger then previous round
            // we advance left, the basic idea here is find which
            // combination of left and right generates the most profit
           if profit > max_profit{
                max_profit = profit;
           }

            
        }
        max_profit
    }
}

File: DSA/leet150/sliding_windown/best_time_to_buy_and_sell_stock_121_easy.c
#include <stdio.h>
#include <stdlib.h>
#include <stdio.h>
int maxProfit(int* prices, int pricesSize){

  int cheapest = prices[0];
  int max_profit = 0;

  // find cheapes
  for(int i = 0; i < pricesSize; ++i){
    
    if(prices[i] < cheapest)
    {
      cheapest = prices[i];
    }
    else // check for max profit
    {
      int profit = prices[i] - cheapest;

      if (profit > max_profit) {
        max_profit = profit;
      }
    }
  }
  return max_profit;

}